src/protocol/context.rs

```diff
@@ -129,13 +148,41 @@ impl<'a, F: Field, SS: SecretSharing<F>> ProtocolContext<'a, SS, F> {
 /// Implementation to upgrade semi-honest context to malicious. Only works for replicated secret
 /// sharing because it is not known yet how to do it for any other type of secret sharing.
 impl<'a, F: Field> ProtocolContext<'a, Replicated<F>, F> {
-    #[must_use]
-    pub fn upgrade_to_malicious(
+    ///
+    /// # Errors
+    /// Performs a multiplication. If that throws an error, so will this
+    ///
+    pub async fn upgrade_to_malicious(
         self,
         accumulator: SecurityValidatorAccumulator<F>,
         r_share: Replicated<F>,
-    ) -> ProtocolContext<'a, MaliciousReplicated<F>, F> {
-        ProtocolContext {
+        record_id: RecordId,
+        input: Replicated<F>,
+    ) -> Result<
+        (
+            ProtocolContext<'a, MaliciousReplicated<F>, F>,
+            MaliciousReplicated<F>,
+        ),
+        Error,
+    > {
```

Comment:
I've been thinking about this and maybe what we really need is a wrapper type being returned here.

What we really need is to pass the accumulator and `r` to the context to get a malicious context.  But we also want to ensure a few other things:

1. The inputs are passed through something to turn them from `Replicated` to `MaliciousReplicated`.  This needs to happen for multiple inputs, but we need just one context from that.  Whatever thing that is needs to have access to a semi-honest context (for multiplication) and the share of `r`.

2. The validation needs to occur at the end of the processing.  This either requires that we pass a malicious context (which is weird, because any would do there) but it would really be better as an object that is consumed.  The return value here could be that thing.

The result would look something like:

```rust
async fn full_protocol(ctx: SemiHonestContext<...>, a: &Replicated<F>, b: &Replicated<F>) -> Replicated<F> {
    // The upgrade function can generate the accumulator and shares of `r`.
    // The return value holds everything needed.
    let wrapper = ctx.upgrade();

    let a = wrapper.upgrade_share(RecordId::from(0), a); // You need to do this for all inputs.
    let b = wrapper.upgrade_share(RecordId::from(1), b);

    // Execute the protocol
    let result = wrapper.context().narrow(&Step::Multiply).multiply(a, b).await?;

    // Validate the values and release the answer.
    wrapper.validate(result).await?
}
```

Something to note here is that the `MaliciousInputValidationStep` implementation of `SubStep` can be extended here.  You currently have: 

* multiply by `r`
* randomness for each input

But you probably also need:

* a narrowing for the malicious protocol execution
* one or more narrowing for the validation steps

Then you have a clean hierarchy of steps, rather than having different types of `SubStep` coexist at the same level of the tree (which will be a problem later).